src/NUnitFramework/framework/Internal/Commands/SetUpTearDownItem.cs

```diff
@@ -84,12 +89,32 @@ public void RunTearDown(TestExecutionContext context)
                     // run the teardowns in reverse order to provide consistency.
                     int index = _tearDownMethods.Count;
                     while (--index >= 0)
-                        Reflect.InvokeMethod(_tearDownMethods[index], _tearDownMethods[index].IsStatic ? null : context.TestObject);
+                    {
+                        var tearDownMethod = _tearDownMethods[index];
+#if NET_4_0 || NET_4_5
+                        if (AsyncInvocationRegion.IsAsyncOperation(tearDownMethod))
+                            RunAsyncMethod(tearDownMethod, context);
+                        else
+#endif
+                            RunNonAsyncMethod(tearDownMethod, context);
```

Comment:
You could extract this logic into a single method called for both setup and teardown.